[JSC] Introduce the "Greedy" Register Allocator for B3 Air

dhecht opened the pull request:
#### f74305d8e5507b0e3da7e4dfc9b3dd3476cb8c6b
<pre>
[JSC] Introduce the &quot;Greedy&quot; Register Allocator for B3 Air
<a href="https://bugs.webkit.org/show_bug.cgi?id=287663">https://bugs.webkit.org/show_bug.cgi?id=287663</a>
<a href="https://rdar.apple.com/144792183">rdar://144792183</a>

Reviewed by Yusuke Suzuki.

Currently disabled but functional.

See the comment at the top of AirAllocateRegistersByGraphColoring.cpp for
a high level view of the design.

Currently B3 Air uses four different register allocator variants:
1. GenerateAndAllocateRegisters allocator: used when optLevel == 0
2. Linear scan: used when optLevel == 1 or there are more tmps than
   Options::maximumTmpsForGraphColoring(), and SIMD is not used by the procedure.
3. IRC graph coloring: used by OMG on ARM and x86, and JS on x86
4. Briggs graph coloring: used by JS on ARM

It&apos;d be nice to have a single register allocator for FTL/OMG that works well
in all configurations. Additionally, the graph coloring allocator can be time
and memory consuming when there are lots of tmps and interference.

This register allocator is inspired by LLVM&apos;s Greedy register allocator and
may work across these configurations with less overhead.

The eventual goal is to enable this allocator by default and phase out and
remove the other register allocators.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
(JSC::B3::Air::allocateRegistersByGraphColoring):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp: Added.
(JSC::B3::Air::Greedy::verbose):
(JSC::B3::Air::Greedy::LiveRange::validate):
(JSC::B3::Air::Greedy::LiveRange::prepend):
(JSC::B3::Air::Greedy::LiveRange::append):
(JSC::B3::Air::Greedy::LiveRange::crossBasicBlockBoundary):
(JSC::B3::Air::Greedy::LiveRange::isLocal const):
(JSC::B3::Air::Greedy::LiveRange::intervals const):
(JSC::B3::Air::Greedy::LiveRange::size):
(JSC::B3::Air::Greedy::LiveRange::overlaps):
(JSC::B3::Air::Greedy::LiveRange::merge):
(JSC::B3::Air::Greedy::LiveRange::subtract):
(JSC::B3::Air::Greedy::LiveRange::dump const):
(JSC::B3::Air::Greedy::TmpPriority::pack_priority):
(JSC::B3::Air::Greedy::TmpPriority::TmpPriority):
(JSC::B3::Air::Greedy::TmpPriority::tmp):
(JSC::B3::Air::Greedy::TmpPriority::stage const):
(JSC::B3::Air::Greedy::TmpPriority::dump const):
(JSC::B3::Air::Greedy::TmpPriority::isHigherPriority):
(JSC::B3::Air::Greedy::RegisterRange::AllocatedInterval::operator&lt; const):
(JSC::B3::Air::Greedy::RegisterRange::AllocatedInterval::dump const):
(JSC::B3::Air::Greedy::RegisterRange::add):
(JSC::B3::Air::Greedy::RegisterRange::evict):
(JSC::B3::Air::Greedy::RegisterRange::hasConflict):
(JSC::B3::Air::Greedy::RegisterRange::forEachConflict):
(JSC::B3::Air::Greedy::RegisterRange::dump const):
(JSC::B3::Air::Greedy::RegisterRange::findFirstIntervalEndingAfter):
(JSC::B3::Air::Greedy::TmpData::CoalescableWith::dump const):
(JSC::B3::Air::Greedy::TmpData::dump const):
(JSC::B3::Air::Greedy::TmpData::isGroup):
(JSC::B3::Air::Greedy::TmpData::validate):
(JSC::B3::Air::Greedy::SplitMetadata::dump const):
(JSC::B3::Air::Greedy::GreedyAllocator::GreedyAllocator):
(JSC::B3::Air::Greedy::GreedyAllocator::run):
(JSC::B3::Air::Greedy::GreedyAllocator::buildRegisterSets):
(JSC::B3::Air::Greedy::GreedyAllocator::buildIndices):
(JSC::B3::Air::Greedy::GreedyAllocator::findBlockContainingPoint):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfHead):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfTail):
(JSC::B3::Air::Greedy::GreedyAllocator::instIndex):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfEarly):
(JSC::B3::Air::Greedy::GreedyAllocator::earlyInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::lateInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::earlyAndLateInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::interval):
(JSC::B3::Air::Greedy::GreedyAllocator::intervalForSpill):
(JSC::B3::Air::Greedy::GreedyAllocator::groupForTmp):
(JSC::B3::Air::Greedy::GreedyAllocator::assignedReg):
(JSC::B3::Air::Greedy::GreedyAllocator::spillSlot):
(JSC::B3::Air::Greedy::GreedyAllocator::adjustedBlockFrequency):
(JSC::B3::Air::Greedy::GreedyAllocator::validateAssignments):
(JSC::B3::Air::Greedy::GreedyAllocator::buildLiveRanges):
(JSC::B3::Air::Greedy::GreedyAllocator::forEachTmpInGroup):
(JSC::B3::Air::Greedy::GreedyAllocator::finalizeGroups):
(JSC::B3::Air::Greedy::GreedyAllocator::initSpillCosts):
(JSC::B3::Air::Greedy::GreedyAllocator::newTmp):
(JSC::B3::Air::Greedy::GreedyAllocator::addSpillTmpWithInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::dumpRegRanges):
(JSC::B3::Air::Greedy::GreedyAllocator::setStageAndEnqueue):
(JSC::B3::Air::Greedy::GreedyAllocator::allocateRegisters):
(JSC::B3::Air::Greedy::GreedyAllocator::tryAllocate):
(JSC::B3::Air::Greedy::GreedyAllocator::tryEvict):
(JSC::B3::Air::Greedy::GreedyAllocator::assign):
(JSC::B3::Air::Greedy::GreedyAllocator::evict):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplit):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplitGroup):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplitAroundClobbers):
(JSC::B3::Air::Greedy::GreedyAllocator::stackSlotMinimumWidth):
(JSC::B3::Air::Greedy::GreedyAllocator::spill):
(JSC::B3::Air::Greedy::GreedyAllocator::queueContainsOnlySpills):
(JSC::B3::Air::Greedy::GreedyAllocator::moveOpcode):
(JSC::B3::Air::Greedy::GreedyAllocator::emitSpillCodeAndEnqueueNewTmps):
(JSC::B3::Air::Greedy::GreedyAllocator::insertFixupCode):
(JSC::B3::Air::Greedy::GreedyAllocator::mayBeCoalescable):
(JSC::B3::Air::Greedy::GreedyAllocator::assignRegisters):
(JSC::B3::Air::allocateRegistersByGreedy):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.h: Added.
* Source/JavaScriptCore/b3/air/AirCode.h:
(JSC::B3::Air::Code::forEachTmp):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::ensureSize):
(JSC::B3::Air::TmpWidth::setWidths):
* Source/JavaScriptCore/b3/air/AirTmpWidth.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/290639@main">https://commits.webkit.org/290639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/428efb43cc8ce969ef0dba1a3749b152801afcdc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90691 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/10228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45637 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/41477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18545 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/95706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/41477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93692 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/40606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/83506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37694 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/89471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/17886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/97536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/18145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78074 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19259 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/21078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/17895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111960 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/17634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/21090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/19418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
```diff
@@ -171,6 +171,26 @@ void TmpWidth::recompute(Code& code)
+    ASSERT(tmp.bank() == bank);
+    auto index = AbsoluteTmpMapper<bank>::absoluteIndex(tmp);
+    auto& bankWidthsVector = widthsVector(bank);
+    if (index >= bankWidthsVector.size())
+        bankWidthsVector.resize(index + 1);
```
If you come into this check multiple times with indices n, n+1, n+2, ..., n+m, you'll end up resizing & potentially reallocating the vector m times. Perhaps instead either push_back up to index or manually do something like `resize(index * some_growth_factor)`?

dhecht replied:
I'm pretty sure WTF::Vector already adds headroom when resizing -- see `Vector::expandCapacity` and `Malloc::nextCapacity`.

Achierius replied:
I see -- I was assuming it had the same behavior in that regard as std::vector does.

Achierius commented on Source/JavaScriptCore/b3/air/AirGenerate.cpp:
```diff
@@ -132,7 +141,10 @@ void prepareForGeneration(Code& code)
         // Register allocation for all the Tmps that do not have a corresponding machine
         // register. After this phase, every Tmp has a reg.
-        allocateRegistersByGraphColoring(code);
+        if (Options::airUseGreedyRegAlloc() && !code.usesSIMD())
+            allocateRegistersByGreedy(code);
```
Out of curiosity, do you know approximately what percentage of the time this (i.e. !code.usesSIMD()) is the case?

dhecht replied:
`usesSIMD()` is only true for Wasm functions that use SIMD heuristics, all other cases will use the new allocator. JS2 And SP3 don't have an SIMD use, so the entire benchmarks uses the new allocator, but JS3 will have SIMD IIRC. Of course, at some point we should add this support, but it hasn't been a priority for me yet.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+
+// Multiplier of 0 means split around clobbers at ever opportunity. The higher the multiplier,
+// the less often the split will be applied (i.e. treats splitting as more costly).
+constexpr float splitCostMultiplier = 0.0f;
```
Am I correct in that the only upside of increasing this would be reduced compile time?

dhecht replied:
No, it will affect output code quality most of all. It controls how often it splits live ranges around clobbers (e.g. calls). I'll be increasing this based on benchmark results later (before enabling this allocator by default) but for now I had it at 0 to stress the splitting logic (it was the last thing I had added).

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+private:
+    Deque<Interval> m_intervals;
+    size_t m_size { 0 }; // Sum of the distances over m_intervals
+    bool crossesBasicBlockBoundary { false };
```
m_crossesBasicBlockBoundary?

dhecht replied:
Oops fixed.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+            auto index = AbsoluteTmpMapper<bank>::absoluteIndex(tmp);
+            float spillCost = m_useCounts.numWarmUsesAndDefs<bank>(index);
+            if (bank == GP && m_useCounts.isConstDef<GP>(index))
+                spillCost /= 2; // Can rematerialize rather than spill in many cases.
```
Is the assumption here that we'll convert the spill into a rematerialization in a later pass? Which one is that?

dhecht replied:
Yes. See the `tryRematerialize()` lambda during `emitSpillCodeAndEnqueueNewTmps`. (Note this was ported over from the graph allocator.)

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+                    spill(tmp, tmpData);
+                    continue;
+                case Stage::Unspillable:
+                    // Unspillables must have been allocated during tryAllocate or tryEvict.
```
May as well add a dataLogLn here like you have for New/Assigned/Spilled/Coalesced/Replaced

dhecht replied:
Yes, thanks. I'll just fall through to that logging code. These different cases pre-existed that log statement and was how I was distinguishing them, but the log output is good enough to distinguish.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+                        if (m_tmpWidth.defWidth(inst.args[0].tmp()) <= Width32)
+                            inst.kind.opcode = Move;
+                    }
+                }
```
I believe we support RISC-V/LoongArch ports, so there should be a fallback case here

dhecht replied:
Not sure I understand the request. These are just some optimizations (ported from the graph allocator) - they also didn't seem to really impact the benchmarks at all IIRC.

Achierius replied:
I see, didn't realize these were just optimizations.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1925 @@
+                        splitCost += adjustedBlockFrequency(block) * 2;
+                        return IterationStatus::Continue;
+                    }
+                    // Non-clobber conflict exists so splitting won't help.
```
What is a "non-clobber conflict"?

dhecht replied:
I conflict that doesn't involve a clobbered register. As in: the live range overlaps another Tmp's assignment to this register. I'll restate it that way.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
```diff
@@ -2276,6 +2276,9 @@ void allocateRegistersByGraphColoring(Code& code)
     graphColoringRegisterAllocation.run();
+
+    if (traceDebug)
+        dataLog("Code after graph coloring:\n", code);
```
Use `dataLogLnIf`.

dhecht replied:
Fixed here and above

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.h:
```diff
@@ -0,0 +1,42 @@
+#if ENABLE(B3_JIT)
+
+#include "CPU.h"
+#include "Options.h"
```
Maybe we do not need these inclusion.

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+    Greedy::GreedyAllocator allocator(code);
+    allocator.run();
+    if (Greedy::verbose())
+        dataLog("Air after greedy register allocation:\n", code);
```
Let's use `dataLogLnIf`.

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+    RELEASE_ASSERT(!code.usesSIMD());
+    PhaseScope phaseScope(code, "allocateRegistersAndStackByGreedy"_s);
+    if (Greedy::verbose())
+        dataLog("Air before greedy register allocation:\n", code);
```
Let's use `dataLogLnIf`.

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+constexpr unsigned spillStore = 0;
+constexpr unsigned splitMoveTo = 1;
+constexpr unsigned splitMoveFrom = 2;
+constexpr unsigned spillLoad = 3;
```
Put `static` to these variables.

dhecht replied:
yeah, fixed. Originally this was all in an anonymous namespace which implies internal linkage, but then I named the namespace since unified-sources was causing all anonymous namespaces in the same unified source to merge.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+
+typedef size_t Point;
+typedef WTF::Range<Point> Interval;
+
```
Use `using`.

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+        }
+    };
+
+    typedef StdSet<AllocatedInterval> AllocatedIntervalSet;
```
Use `using`

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+            tailPosition = headPosition + 2 * block->size() - 1;
+            m_blockToHeadPoint[block] = headPosition;
+            m_tailPoints[i] = tailPosition;
+            headPosition += 2 * block->size();
```
Maybe we would like to have a constant variable to `2`

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+
+    static size_t instIndex(Point positionOfHead, Point point)
+    {
+        return (point - positionOfHead) / 2;
```
Ditto.

dhecht replied:
Fixed

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -0,0 +1,1924 @@
+
+    Point positionOfTail(BasicBlock* block)
+    {
+        return positionOfHead(block) + block->size() * 2 - 1;
```
Ditto.

dhecht replied:
Fixed

Constellation reviewed (approved):
I reviewed the algorithm and I think this looks good.

dhecht committed:
[JSC] Introduce the "Greedy" Register Allocator for B3 Air
https://bugs.webkit.org/show_bug.cgi?id=287663
rdar://144792183

Reviewed by Yusuke Suzuki.

Currently disabled but functional.

See the comment at the top of AirAllocateRegistersByGraphColoring.cpp for
a high level view of the design.

Currently B3 Air uses four different register allocator variants:
1. GenerateAndAllocateRegisters allocator: used when optLevel == 0
2. Linear scan: used when optLevel == 1 or there are more tmps than
   Options::maximumTmpsForGraphColoring(), and SIMD is not used by the procedure.
3. IRC graph coloring: used by OMG on ARM and x86, and JS on x86
4. Briggs graph coloring: used by JS on ARM

It'd be nice to have a single register allocator for FTL/OMG that works well
in all configurations. Additionally, the graph coloring allocator can be time
and memory consuming when there are lots of tmps and interference.

This register allocator is inspired by LLVM's Greedy register allocator and
may work across these configurations with less overhead.

The eventual goal is to enable this allocator by default and phase out and
remove the other register allocators.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
(JSC::B3::Air::allocateRegistersByGraphColoring):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp: Added.
(JSC::B3::Air::Greedy::verbose):
(JSC::B3::Air::Greedy::LiveRange::validate):
(JSC::B3::Air::Greedy::LiveRange::prepend):
(JSC::B3::Air::Greedy::LiveRange::append):
(JSC::B3::Air::Greedy::LiveRange::crossBasicBlockBoundary):
(JSC::B3::Air::Greedy::LiveRange::isLocal const):
(JSC::B3::Air::Greedy::LiveRange::intervals const):
(JSC::B3::Air::Greedy::LiveRange::size):
(JSC::B3::Air::Greedy::LiveRange::overlaps):
(JSC::B3::Air::Greedy::LiveRange::merge):
(JSC::B3::Air::Greedy::LiveRange::subtract):
(JSC::B3::Air::Greedy::LiveRange::dump const):
(JSC::B3::Air::Greedy::TmpPriority::pack_priority):
(JSC::B3::Air::Greedy::TmpPriority::TmpPriority):
(JSC::B3::Air::Greedy::TmpPriority::tmp):
(JSC::B3::Air::Greedy::TmpPriority::stage const):
(JSC::B3::Air::Greedy::TmpPriority::dump const):
(JSC::B3::Air::Greedy::TmpPriority::isHigherPriority):
(JSC::B3::Air::Greedy::RegisterRange::AllocatedInterval::operator< const):
(JSC::B3::Air::Greedy::RegisterRange::AllocatedInterval::dump const):
(JSC::B3::Air::Greedy::RegisterRange::add):
(JSC::B3::Air::Greedy::RegisterRange::evict):
(JSC::B3::Air::Greedy::RegisterRange::hasConflict):
(JSC::B3::Air::Greedy::RegisterRange::forEachConflict):
(JSC::B3::Air::Greedy::RegisterRange::dump const):
(JSC::B3::Air::Greedy::RegisterRange::findFirstIntervalEndingAfter):
(JSC::B3::Air::Greedy::TmpData::CoalescableWith::dump const):
(JSC::B3::Air::Greedy::TmpData::dump const):
(JSC::B3::Air::Greedy::TmpData::isGroup):
(JSC::B3::Air::Greedy::TmpData::validate):
(JSC::B3::Air::Greedy::SplitMetadata::dump const):
(JSC::B3::Air::Greedy::GreedyAllocator::GreedyAllocator):
(JSC::B3::Air::Greedy::GreedyAllocator::run):
(JSC::B3::Air::Greedy::GreedyAllocator::buildRegisterSets):
(JSC::B3::Air::Greedy::GreedyAllocator::buildIndices):
(JSC::B3::Air::Greedy::GreedyAllocator::findBlockContainingPoint):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfHead):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfTail):
(JSC::B3::Air::Greedy::GreedyAllocator::instIndex):
(JSC::B3::Air::Greedy::GreedyAllocator::positionOfEarly):
(JSC::B3::Air::Greedy::GreedyAllocator::earlyInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::lateInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::earlyAndLateInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::interval):
(JSC::B3::Air::Greedy::GreedyAllocator::intervalForSpill):
(JSC::B3::Air::Greedy::GreedyAllocator::groupForTmp):
(JSC::B3::Air::Greedy::GreedyAllocator::assignedReg):
(JSC::B3::Air::Greedy::GreedyAllocator::spillSlot):
(JSC::B3::Air::Greedy::GreedyAllocator::adjustedBlockFrequency):
(JSC::B3::Air::Greedy::GreedyAllocator::validateAssignments):
(JSC::B3::Air::Greedy::GreedyAllocator::buildLiveRanges):
(JSC::B3::Air::Greedy::GreedyAllocator::forEachTmpInGroup):
(JSC::B3::Air::Greedy::GreedyAllocator::finalizeGroups):
(JSC::B3::Air::Greedy::GreedyAllocator::initSpillCosts):
(JSC::B3::Air::Greedy::GreedyAllocator::newTmp):
(JSC::B3::Air::Greedy::GreedyAllocator::addSpillTmpWithInterval):
(JSC::B3::Air::Greedy::GreedyAllocator::dumpRegRanges):
(JSC::B3::Air::Greedy::GreedyAllocator::setStageAndEnqueue):
(JSC::B3::Air::Greedy::GreedyAllocator::allocateRegisters):
(JSC::B3::Air::Greedy::GreedyAllocator::tryAllocate):
(JSC::B3::Air::Greedy::GreedyAllocator::tryEvict):
(JSC::B3::Air::Greedy::GreedyAllocator::assign):
(JSC::B3::Air::Greedy::GreedyAllocator::evict):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplit):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplitGroup):
(JSC::B3::Air::Greedy::GreedyAllocator::trySplitAroundClobbers):
(JSC::B3::Air::Greedy::GreedyAllocator::stackSlotMinimumWidth):
(JSC::B3::Air::Greedy::GreedyAllocator::spill):
(JSC::B3::Air::Greedy::GreedyAllocator::queueContainsOnlySpills):
(JSC::B3::Air::Greedy::GreedyAllocator::moveOpcode):
(JSC::B3::Air::Greedy::GreedyAllocator::emitSpillCodeAndEnqueueNewTmps):
(JSC::B3::Air::Greedy::GreedyAllocator::insertFixupCode):
(JSC::B3::Air::Greedy::GreedyAllocator::mayBeCoalescable):
(JSC::B3::Air::Greedy::GreedyAllocator::assignRegisters):
(JSC::B3::Air::allocateRegistersByGreedy):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.h: Added.
* Source/JavaScriptCore/b3/air/AirCode.h:
(JSC::B3::Air::Code::forEachTmp):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::ensureSize):
(JSC::B3::Air::TmpWidth::setWidths):
* Source/JavaScriptCore/b3/air/AirTmpWidth.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/290639@main
